src/bin/server.rs

```diff
@@ -56,37 +60,124 @@ fn main() -> anyhow::Result<()> {
+        let listener = TcpListener::bind((app.config.ip, app.config.port)).await?;
+
+        let addr = listener.local_addr()?;
+
+        // Do not change this line! Removing the line or changing its contents in any way will break
+        // the test suite :)
+        info!("Listening at http://{addr}");
+
+        // Create a watch channel to track tasks that are handling connections and wait for them to
+        // complete.
+        let (close_tx, close_rx) = watch::channel(());
+
+        // Continuously accept new connections.
+        loop {
+            let (socket, remote_addr) = tokio::select! {
+                // Either accept a new connection...
+                result = listener.accept() => {
+                    result.unwrap()
+                }
+                // ...or wait to receive a shutdown signal and stop the accept loop.
+                _ = shutdown_signal() => {
+                    debug!("shutdown signal received, not accepting new connections");
+                    break;
+                }
             };
 
-            info!("Starting graceful shutdown");
-        });
+            debug!("connection {remote_addr} accepted");
+
+            // We don't need to call `poll_ready` because `IntoMakeServiceWithConnectInfo` is always
+            // ready.
+            let tower_service = make_service.call(remote_addr).await.unwrap();
+
+            // Clone the watch receiver and move it into the task.
+            let close_rx = close_rx.clone();
+
+            // Spawn a task to handle the connection. That way we can serve multiple connections
+            // concurrently.
+            tokio::spawn(async move {
+                // Hyper has its own `AsyncRead` and `AsyncWrite` traits and doesn't use tokio.
+                // `TokioIo` converts between them.
+                let socket = TokioIo::new(socket);
+
+                // Hyper also has its own `Service` trait and doesn't use tower. We can use
+                // `hyper::service::service_fn` to create a hyper `Service` that calls our app through
+                // `tower::Service::call`.
+                let hyper_service =
+                    hyper::service::service_fn(move |request: Request<Incoming>| {
+                        // We have to clone `tower_service` because hyper's `Service` uses `&self` whereas
+                        // tower's `Service` requires `&mut self`.
+                        //
+                        // We don't need to call `poll_ready` since `Router` is always ready.
+                        tower_service.clone().call(request.map(axum::body::Body::new))
+                    });
+
+                // `hyper_util::server::conn::auto::Builder` supports both http1 and http2 but doesn't
+                // support graceful so we have to use hyper directly and unfortunately pick between
+                // http1 and http2.
+                let conn = hyper::server::conn::http1::Builder::new()
```

Comment:
I was wondering if this was going to be a regression, but it looks like we only support HTTP 1.1 today on the actual crates.io instance anyway, so this shouldn't change anything.